simapp/codec/codec.go

```diff
@@ -0,0 +1,121 @@
+package codec
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	authexported "github.com/cosmos/cosmos-sdk/x/auth/exported"
+	"github.com/cosmos/cosmos-sdk/x/auth/vesting"
+	"github.com/cosmos/cosmos-sdk/x/supply"
+	"github.com/cosmos/cosmos-sdk/x/supply/exported"
+)
+
+var (
+	_ auth.Codec   = (*Codec)(nil)
+	_ supply.Codec = (*Codec)(nil)
+)
+
+// AppCodec defines the application-level codec. This codec contains all the
+// required module-specific codecs that are to be provided upon initialization.
+type Codec struct {
+	codec.Marshaler
+
+	// Keep reference to the amino codec to allow backwards compatibility along
+	// with type, and interface registration.
+	amino *codec.Codec
+}
+
+func NewAppCodec(amino *codec.Codec) *Codec {
+	return &Codec{Marshaler: codec.NewHybridCodec(amino), amino: amino}
+}
+
+// MarshalAccount marshals an Account interface. If the given type implements
+// the Marshaler interface, it is treated as a Proto-defined message and
+// serialized that way. Otherwise, it falls back on the internal Amino codec.
+func (c *Codec) MarshalAccount(accI authexported.Account) ([]byte, error) {
+	acc := &Account{}
+	acc.SetAccount(accI)
```

Comment:
This should return an `error` if it fails and that should be passed up to the caller. The simapp `Account` `oneof` may not be able to handle all the cases.

Reply:
ACK